include/deal.II/base/mpi_noncontiguous_partitioner.templates.h

```diff
@@ -321,17 +323,25 @@ namespace Utilities
           AssertThrowMPI(ierr);
         }
 
-      auto src_iterator = src.begin();
-
       // post send
       for (types::global_dof_index i = 0, k = 0; i < send_ranks.size(); i++)
         {
           // collect data to be send
+          AssertIndexRange(i + 1, send_ptr.size());
           for (types::global_dof_index j = send_ptr[i]; j < send_ptr[i + 1];
                j++)
-            buffers[j] = src_iterator[send_indices[k++]];
+            {
+              AssertIndexRange(k, send_indices.size());
+              buffers[j] = src[send_indices[k]];
+              ++k;
+            }
 
           // send data
+          Assert((send_ptr[i] < buffers.size()) ||
+                   (send_ptr[i] == buffers.size() &&
+                    send_ptr[i + 1] == send_ptr[i]),
+                 ExcMessage("The input buffer doesn't contain enough entries"));
+          AssertIndexRange(i + 1, send_ptr.size());
```

Comment:
```suggestion
          AssertIndexRange(send_ranks.size(), send_ptr.size());
```
and this is then a duplicate of the one in l.330.